### ISSUE TITLE
Worker should run as user

### ISSUE THREAD

**Issue by srossross** [Anaconda-Server/anaconda-build#84]
The worker should run as root and should execute the build script as another lesser user .e.g. `anaconda-worker`.

The worker should kill all processes for the `anaconda-worker` before a job starts and clean the home directory by removing it and replacing it with `/etc/worker-skel` on unix.

I'm not sure how to accomplish this on win32 machines

**Comment by stephenakearns** [comment]
quick follow up note - Michael Sarahan was suggested as a possible Windows
Dev who may be able to help out with support for Win32/64

On Wed, Sep 2, 2015 at 1:25 PM, Sean <notifications@github.com> wrote:

> The worker should run as root and should execute the build script as
> another lesser user .e.g. anaconda-worker.
>
> The worker should kill all processes for the anaconda-worker before a job
> starts and clean the home directory by removing it and replacing it with
> /etc/worker-skel on unix.
>
> I'm not sure how to accomplish this on win32 machines
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Anaconda-Server/anaconda-build/issues/84>.
>



-- 
*Stephen Kearns*
*Product Manager*

**Comment by bkreider** [comment]
Can this be configurable?  Some places don't want things running things as root (or delegating access to users that might have a lot of privileges).  You could have multiple system users -- worker1-workerN -- that only have write permission to their home directories.  

Is there a reason the build workers need to masquerade as a real person when building?

**Comment by bkreider** [comment]
@mrduane You might have some good customer feedback for this issue since it would require customers to run it as root.

**Comment by srossross** [comment]
@bkreider the goal of this is to better isolate builds,  do you have any recommendations of other directions we could take?

> Can this be configurable?

Yes, I was thinking of adding a `--build-user anaconda-worker` argument to enable this.

**Comment by bkreider** [comment]
I don't think the build users need to be actual users or tied to the AS users at all.  I think if we can have it work without running as root and without using sudo, it will be much easier to sell.  If it needs sudo, if we can lock it down to a very few commands that need to run as sudo, that is very different than running as root.  

Without thinking about this very long, here is a  hacky way of doing it.  I'm sure we can come up with something better:
1.  Admin creates 10 worker users and 10 ssh keys for them.
2.  SSH private keys are given to the worker service.  
3.  The worker service runs:  ssh worker1@localhost `build command`  # this pulls the build script from somewhere read-only and the artifacts are readable by the worker service.
4.  The worker service runs:  ssh worker1@localhost `<clean environment command>`  

That's off the top of my head, so I think we can come up with something more elegant that doesn't need the build daemon to run as root.

**Comment by srossross** [comment]
I like this idea, an ssh option would provide more flexibility in other use cases as well. Do you think this solution would be robust enough to use in production?  What do you think the best `clean environment command ` would be? Would we use scp to copy the build script to `worker1@localhost `?

**Comment by srossross** [comment]
> 3. The worker service runs: ssh worker1@localhost build command # this pulls the build script from somewhere read-only and the artifacts are readable by the worker service.

This would make the assumption that the script was globally accessible. Maybe it should be stored on the anaconda-server and fetched with curl?

e.g. `worker1@localhost "curl anaconda.org/build-scripts/script?signature=SECURE > builds/user/package/script.sh && bash builds/user/package/script.sh"`

**Comment by bkreider** [comment]
I think it is pretty solid for not using sudo/su/root/suid.  I was thinking of the enterprise use case.  In cloud, you would want to create the users from scratch each time to make sure they don't mess with the .bash_profiile, .bashrc in a way that clean isn't able to deal with.  There's way to lock it down further, but it is easier to create new users.

For enterprise, I think it is pretty strong.  If they require stronger security, we could give an option to setup projects to use a dedicated user account (that requires more admin work but that is the tradeoff). 

> This would make the assumption that the script was globally accessible. Maybe it should be stored on the anaconda-server and fetched with curl?

It could be pull or push.  I don't know which is better.  The nice thing about pull is that it won't matter if they turn off scp in the ssh server.

**Comment by PeterDSteinberg** [comment]
I am closing this issue even though the PR is not merged yet.  Everything in terms of functionality and testing has been done on su worker, but there are two bugs in conda that are preventing final merge.  I will raise a separate issue for those two bugs and the related followup testing on anaconda-build's su worker after those conda bugs are fixed.  See #126

**Comment by stephenakearns** [comment]
Hi @PeterDSteinberg - just commented on #126 - yes, let's revisit - did the conda fixes resolve your issues? (In another case, there have been secondary issues that emerged after the fix)

**Comment by PeterDSteinberg** [comment]
I just changed the number of story points for this issue from 5 to 13 points.  I changed it because:
 * There were a number of bugs discovered in dependencies that occurred when the dependencies were used in less common use cases (e.g. root python conda-build permissions for other users)
 * A poorly planned test (written by me) in this PR caused build workers that were not su workers to hang indefinitely.  The test ran fine in ```python -m unittest discover``` locally without hanging, but in the build worker setting (with ```anaconda worker run``` not ```anaconda worker su_run```) the build did not finish or get interrupted by iotimeout because of running subprocesses without wait or kill statements. 
 * The hanging builds also overlapped in time with logging problems for the build workers (October), making it harder to figure out.
